Add PolynomialConstraint to solvers

ggould-tri opened the pull request:
In preparation for parameter estimation (for system identification):  Add polynomial constraints to Optimizers.

This is a minimum-viable-change:  It implements polynomial constraints as generic constraints (even though some of them are linear/affine and all could be handled specially by the right solver), and it makes no attempt to fix bugs in the nlopt wrapper that limit their usefulness.  Those matters will be handled in their own due time.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/2215)
<!-- Reviewable:end -->

liangfok commented:
Done with initial review.

---

Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 21 unresolved discussions, some commit checks failed.

---

*[drake/solvers/Constraint.h, line 100 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGUUPFkbV_znbc2slwh:-KGUUPFkbV_znbc2slwi:-1537298934) ([raw file](https://github.com/robotlocomotion/drake/blob/16d0e183c45043c162f10e81563f15edb17f57ec/drake/solvers/Constraint.h#L100)):*
Consider changing:

> lb <= P(x, y...) <= ub  (for P a multivariate polynomial in x, y...)

to be:

> lb <= P(x, y...) <= ub where P is a multivariate polynomial in x, y...

---

*[drake/solvers/Constraint.h, line 104 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGUWHO__ORfA88SGbrU:-KGUWHOaQ61TzXI3Fn_U:1842966468) ([raw file](https://github.com/robotlocomotion/drake/blob/16d0e183c45043c162f10e81563f15edb17f57ec/drake/solvers/Constraint.h#L104)):*
", and so" --> ". Thus, 

---

*[drake/solvers/Constraint.h, line 106 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGUV25oX8Uyi6pX4AEm:-KGUV25oX8Uyi6pX4AEn:352016758) ([raw file](https://github.com/robotlocomotion/drake/blob/16d0e183c45043c162f10e81563f15edb17f57ec/drake/solvers/Constraint.h#L106)):*
Consider changing:

> caller must provide a list of Polynomial::VarType to allow bindings to map the individual scalar elements of the given VariableList.

to be:

> caller must provide a list of Polynomial::VarType variables that specify the bindings between the variables in the polynomial and those in the VariableList.

---

*[drake/solvers/Constraint.h, line 114 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGUXvqFYF9wJLh5uIv6:-KGUXvqFYF9wJLh5uIv7:-250208817) ([raw file](https://github.com/robotlocomotion/drake/blob/16d0e183c45043c162f10e81563f15edb17f57ec/drake/solvers/Constraint.h#L114)):*
The the first parameter supplied to `Constraint` is meaningless to someone who hasn't studied `Constraint`'s API. Thus, I recommend replacing it with:

```
static const int kNumConstraints = 1;

...

Constraint(kNumConstraints,...
```
?

---

*[drake/solvers/Constraint.h, line 117 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGUZ00nMPYY4-B85_up:-KGUZ00nMPYY4-B85_uq:415428317) ([raw file](https://github.com/robotlocomotion/drake/blob/16d0e183c45043c162f10e81563f15edb17f57ec/drake/solvers/Constraint.h#L117)):*
I recommend adding a blank line between the constructor and destructor.

---

*[drake/solvers/Constraint.h, line 120 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGUZAzSDzUdOaQkuAqn:-KGUZAzSDzUdOaQkuAqo:-145161360) ([raw file](https://github.com/robotlocomotion/drake/blob/16d0e183c45043c162f10e81563f15edb17f57ec/drake/solvers/Constraint.h#L120)):*
Add Doxygen description of this method.

---

*[drake/solvers/Constraint.h, line 121 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGU_7JQuLYImSQ3ndTO:-KGU_7JQuLYImSQ3ndTP:145130111) ([raw file](https://github.com/robotlocomotion/drake/blob/16d0e183c45043c162f10e81563f15edb17f57ec/drake/solvers/Constraint.h#L121)):*
Is it possible to avoid instantiating this map every time `eval()` is called?  Perhaps using [clear()](http://en.cppreference.com/w/cpp/container/map/clear)? 

---

*[drake/solvers/Constraint.h, line 129 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGU_aZqHueLs2XaZSQo:-KGU_aZqHueLs2XaZSQp:-145161360) ([raw file](https://github.com/robotlocomotion/drake/blob/16d0e183c45043c162f10e81563f15edb17f57ec/drake/solvers/Constraint.h#L129)):*
Add Doxygen description of this method.

---

*[drake/solvers/Optimization.h, line 558 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGU_y3MN1ger5ALTcYi:-KGU_y3MN1ger5ALTcYj:480601245) ([raw file](https://github.com/robotlocomotion/drake/blob/16d0e183c45043c162f10e81563f15edb17f57ec/drake/solvers/Optimization.h#L558)):*
"adds" --> "Adds"

---

*[drake/solvers/Optimization.h, line 559 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGU_ztYkMVvCPh5cCT5:-KGU_ztYkMVvCPh5cCT6:-8335107) ([raw file](https://github.com/robotlocomotion/drake/blob/16d0e183c45043c162f10e81563f15edb17f57ec/drake/solvers/Optimization.h#L559)):*
Missing period at end of sentence.

Is the subset of the decision variables equal to poly_vars? Where is the entire set of decision variables defined? In the polynomial?

---

*[drake/solvers/Optimization.h, line 569 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGUaKKGNo64iDhkAEcA:-KGUaKKGNo64iDhkAEcB:773564327) ([raw file](https://github.com/robotlocomotion/drake/blob/16d0e183c45043c162f10e81563f15edb17f57ec/drake/solvers/Optimization.h#L569)):*
What does "LC" stand for?

---

*[drake/solvers/Optimization.h, line 583 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGUa_dkBaT0dBGCG7YU:-KGUa_dkBaT0dBGCG7YV:480601245) ([raw file](https://github.com/robotlocomotion/drake/blob/16d0e183c45043c162f10e81563f15edb17f57ec/drake/solvers/Optimization.h#L583)):*
"adds" --> "Adds"

---

*[drake/solvers/Optimization.h, line 584 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGUaboxKkzdQYZQKTbk:-KGUaboxKkzdQYZQKTbl:1292687508) ([raw file](https://github.com/robotlocomotion/drake/blob/16d0e183c45043c162f10e81563f15edb17f57ec/drake/solvers/Optimization.h#L584)):*
Missing period at end of sentence.

---

*[drake/solvers/test/testOptimizationProblem.cpp, line 581 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGUbHq1_LGLTcfD2oK1:-KGUb2erK2TJ5YjQUhFN:-596061632) ([raw file](https://github.com/robotlocomotion/drake/blob/16d0e183c45043c162f10e81563f15edb17f57ec/drake/solvers/test/testOptimizationProblem.cpp#L581)):*
"has two minima" --> "has two minima, one at Foo and the other at Bar"

---

*[drake/solvers/test/testOptimizationProblem.cpp, line 582 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGUbAFWjmAa-dYrW2QP:-KGUbAFWjmAa-dYrW2QQ:-995453870) ([raw file](https://github.com/robotlocomotion/drake/blob/16d0e183c45043c162f10e81563f15edb17f57ec/drake/solvers/test/testOptimizationProblem.cpp#L582)):*
What does "require the negative one" mean? Do you mean "expect the negative one?"

---

*[drake/util/Polynomial.h, line 205 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGUbZjmp9XxfwL9PPSu:-KGUbZjmp9XxfwL9PPSv:159538575) ([raw file](https://github.com/robotlocomotion/drake/blob/16d0e183c45043c162f10e81563f15edb17f57ec/drake/util/Polynomial.h#L205)):*
"eg" --> "e.g."

---

*[drake/util/Polynomial.h, line 215 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGUblNClyUA2edcJjON:-KGUblNClyUA2edcJjOO:1291002751) ([raw file](https://github.com/robotlocomotion/drake/blob/16d0e183c45043c162f10e81563f15edb17f57ec/drake/util/Polynomial.h#L215)):*
I recommend separating the two `pow()` methods with a blank line.

---

*[drake/util/Polynomial.h, line 227 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGUbv-An4G5C3iNQ9KL:-KGUbv-An4G5C3iNQ9KM:-475227472) ([raw file](https://github.com/robotlocomotion/drake/blob/16d0e183c45043c162f10e81563f15edb17f57ec/drake/util/Polynomial.h#L227)):*
"contains values" --> "contains variables"

---

*[drake/util/Polynomial.h, line 252 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGUcIgofD3oGiW8qgOw:-KGUcIgofD3oGiW8qgOx:250617404) ([raw file](https://github.com/robotlocomotion/drake/blob/16d0e183c45043c162f10e81563f15edb17f57ec/drake/util/Polynomial.h#L252)):*
extra space after "of"

---

*[drake/util/Polynomial.h, line 259 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGUcRqTpUX9tW_8hj5u:-KGUcRqTpUX9tW_8hj5v:-635472613) ([raw file](https://github.com/robotlocomotion/drake/blob/16d0e183c45043c162f10e81563f15edb17f57ec/drake/util/Polynomial.h#L259)):*
"another:  " --> "another. "

---

*[drake/util/Polynomial.h, line 260 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGUcWl35mwDaa5eiGiI:-KGUcWl35mwDaa5eiGiJ:-324513893) ([raw file](https://github.com/robotlocomotion/drake/blob/16d0e183c45043c162f10e81563f15edb17f57ec/drake/util/Polynomial.h#L260)):*
"whatever" --> "what"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-:-KGUcc3LMzn-zI1IS6ZH:2062893049)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Fixes in r2.  Thanks!
r2 also adds test disabling logic for windows, as the windows CI builder has no solver for generic constraints.

---

Review status: 0 of 4 files reviewed at latest revision, 22 unresolved discussions.

---

*[drake/solvers/Constraint.h, line 106 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGUV25oX8Uyi6pX4AEm:-KGX3SejKKjCHDnAhr8H:-445422632) ([raw file](https://github.com/robotlocomotion/drake/blob/16d0e183c45043c162f10e81563f15edb17f57ec/drake/solvers/Constraint.h#L106)):*
I worded this slightly differently because Binding refers to a class in Optimization, not the general notion of identity between variable elements..

---

*[drake/solvers/Constraint.h, line 114 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGUXvqFYF9wJLh5uIv6:-KGX4E4d9xudXayLNqZ1:350205076) ([raw file](https://github.com/robotlocomotion/drake/blob/16d0e183c45043c162f10e81563f15edb17f57ec/drake/solvers/Constraint.h#L114)):*
Also made the corresponding change in `QuadraticConstraint` (from which I copied it `Constraint(1` idiom.

---

*[drake/solvers/Constraint.h, line 117 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGUZ00nMPYY4-B85_up:-KGX4_CFbXj6Mvn-I-o6:-320496696) ([raw file](https://github.com/robotlocomotion/drake/blob/16d0e183c45043c162f10e81563f15edb17f57ec/drake/solvers/Constraint.h#L117)):*
Also made the corresponding change to every single other class in this file.

---

*[drake/solvers/Constraint.h, line 120 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGUZAzSDzUdOaQkuAqn:-KGX4jtVtCJt5npdWC8f:1115135078) ([raw file](https://github.com/robotlocomotion/drake/blob/16d0e183c45043c162f10e81563f15edb17f57ec/drake/solvers/Constraint.h#L120)):*
We do not document any Constraint subclass's `::eval()` method because they override a documented method of the superclass.

---

*[drake/solvers/Constraint.h, line 129 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGU_aZqHueLs2XaZSQo:-KGX5Xxgq_ipHLIS1iQq:1115135078) ([raw file](https://github.com/robotlocomotion/drake/blob/16d0e183c45043c162f10e81563f15edb17f57ec/drake/solvers/Constraint.h#L129)):*
We do not document any Constraint subclass's `::eval()` method because they override a documented method of the superclass.

---

*[drake/solvers/Optimization.h, line 558 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGU_y3MN1ger5ALTcYi:-KGX5sB9xVVjXzODikUw:1052636165) ([raw file](https://github.com/robotlocomotion/drake/blob/16d0e183c45043c162f10e81563f15edb17f57ec/drake/solvers/Optimization.h#L558)):*
Also fixed this on every single other `Add...` method of this class.

---

*[drake/solvers/Optimization.h, line 559 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGU_ztYkMVvCPh5cCT5:-KGX6F5BM81hLjffs12g:979212009) ([raw file](https://github.com/robotlocomotion/drake/blob/16d0e183c45043c162f10e81563f15edb17f57ec/drake/solvers/Optimization.h#L559)):*
Also added periods to most/all of the other doc comments in this file.

The "a subset of the decision variables" is present in every `...vars` overload in this class, and refers to `vars`

---

*[drake/solvers/Optimization.h, line 569 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGUaKKGNo64iDhkAEcA:-KGX6fQY_5b0L1Bfcu9h:1320199319) ([raw file](https://github.com/robotlocomotion/drake/blob/16d0e183c45043c162f10e81563f15edb17f57ec/drake/solvers/Optimization.h#L569)):*
Linear complementarity.  I have expanded it in the comment.

---

*[drake/util/Polynomial.h, line 259 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGUcRqTpUX9tW_8hj5u:-KGX7S8NQp6V4pDYjgsZ:182356525) ([raw file](https://github.com/robotlocomotion/drake/blob/16d0e183c45043c162f10e81563f15edb17f57ec/drake/util/Polynomial.h#L259)):*
Replaced ":" instead with "because" to emphasize that the latter fact implies the former.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-:-KGX81Y7zJM6MKO4lklf:-1667039180)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Thanks for the R2 changes. I just left some more minor comments.

---

Reviewed 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/solvers/Constraint.h, line 106 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGUV25oX8Uyi6pX4AEm:-KGXIxtNhqfdXhbvwTs4:-1757670433) ([raw file](https://github.com/robotlocomotion/drake/blob/16d0e183c45043c162f10e81563f15edb17f57ec/drake/solvers/Constraint.h#L106)):*
Thanks for the changes. When reading the new text, it's still not clear to me that "the Binding" is referring to a Binding class in Optimization. Perhaps add a sentence to the effect of, for more details, see the Binding class within the Optimization class.

---

*[drake/solvers/Optimization.h, line 560 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGXK1yIIlDwqsJujS7Z:-KGXK1yJiNLMKuJCNAUr:315458265) ([raw file](https://github.com/robotlocomotion/drake/blob/c0194da8a8ab4b46c66baffec0a168446957e33a/drake/solvers/Optimization.h#L560)):*
Consider changing:

```
variables.
```

to be:

```
variables, which are defined in parameter vars.
```

---

*[drake/util/Polynomial.h, line 262 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGXLaAl6BOs2xI5efq8:-KGXLaAl6BOs2xI5efq9:2117761853) ([raw file](https://github.com/robotlocomotion/drake/blob/c0194da8a8ab4b46c66baffec0a168446957e33a/drake/util/Polynomial.h#L262)):*
Am I misreading this as saying `var_values` indices can correspond to __both__ variables and vector elements?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-:-KGXM51NUbcWTchU30bg:-2072233625)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/solvers/Constraint.h, line 106 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGUV25oX8Uyi6pX4AEm:-KGXNQPjXErACt-QdwBL:-890532975) ([raw file](https://github.com/robotlocomotion/drake/blob/16d0e183c45043c162f10e81563f15edb17f57ec/drake/solvers/Constraint.h#L106)):*
Added its enclosing namespace, which both clarifies that it is a class and gives an indication of where to look.

---

*[drake/solvers/Optimization.h, line 560 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGXK1yIIlDwqsJujS7Z:-KGXNyj-jJ1eiujMZRfT:2000161824) ([raw file](https://github.com/robotlocomotion/drake/blob/c0194da8a8ab4b46c66baffec0a168446957e33a/drake/solvers/Optimization.h#L560)):*
Updated in every `Add...(...vars)` method of this class

---

*[drake/util/Polynomial.h, line 262 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGXLaAl6BOs2xI5efq8:-KGXOC_mKaRpchjl_IMf:1597815361) ([raw file](https://github.com/robotlocomotion/drake/blob/c0194da8a8ab4b46c66baffec0a168446957e33a/drake/util/Polynomial.h#L262)):*
They can refer to anything that the caller wants -- that is exactly the problem.  I have reworded the comment a bit but ultimately anyone using TaylorVec will eventually have to understand AutoDiff, and a comment in Polynomial is not the place for that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2215)*
<!-- Sent from Reviewable.io -->

liangfok commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-:-KGXSEPaHh5awW8iMUXg:-1868698381)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/solvers/Constraint.h, line 133 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGXW5tiXXZ8j9kThFK7:-KGXW5tiXXZ8j9kThFK8:-1264902853) ([raw file](https://github.com/robotlocomotion/drake/blob/12d3139d847511e9decd138ff584a1f7aa7138b0/drake/solvers/Constraint.h#L133)):*
Would allocating this at object construction time be so bad?  At least it would be thread safe then...  Also we're assuming clear() doesn't free memory and then reallocate when the [] operator is invoked (otherwise just creating the map on the stack would be fine).  Is there actually a guarantee of this?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2215)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/solvers/Constraint.h, line 133 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGXW5tiXXZ8j9kThFK7:-KGXYQnp8xZIvIBjSMKz:385213538) ([raw file](https://github.com/robotlocomotion/drake/blob/12d3139d847511e9decd138ff584a1f7aa7138b0/drake/solvers/Constraint.h#L133)):*
I'm going to let you and @liangfok argue over that one.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2215)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/solvers/Constraint.h, line 133 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGXW5tiXXZ8j9kThFK7:-KGXZVv1yCaW9qLUoBXe:1287741607) ([raw file](https://github.com/robotlocomotion/drake/blob/12d3139d847511e9decd138ff584a1f7aa7138b0/drake/solvers/Constraint.h#L133)):*
I can't see how `static` here is possibly correct.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2215)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/solvers/Constraint.h, line 133 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGXW5tiXXZ8j9kThFK7:-KGXbx594BKzfDDrzAEG:-1356021387) ([raw file](https://github.com/robotlocomotion/drake/blob/12d3139d847511e9decd138ff584a1f7aa7138b0/drake/solvers/Constraint.h#L133)):*
OK for moving it to the constructor and as a member variable of the class (no need for it to be static).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2215)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Review status: 3 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/solvers/Constraint.h, line 133 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGXW5tiXXZ8j9kThFK7:-KGYAF5lsjsbc7jWb-hI:12291164) ([raw file](https://github.com/robotlocomotion/drake/blob/12d3139d847511e9decd138ff584a1f7aa7138b0/drake/solvers/Constraint.h#L133)):*
I have moved the pre-allocated map to a member in r4

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2215)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: 3 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/solvers/Constraint.h, line 133 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGXW5tiXXZ8j9kThFK7:-KGYBjb1DK7PuoWxiCS1:840487005) ([raw file](https://github.com/robotlocomotion/drake/blob/12d3139d847511e9decd138ff584a1f7aa7138b0/drake/solvers/Constraint.h#L133)):*
:ok_hand:

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2215)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 3 of 4 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-:-KGYG9k0Xs7MoQrwh_wy:1417088721)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Windows failures look spurious.  I can't figure out what is even the error in the mac build.

liangfok commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2215)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
> Windows failures look spurious.

Yes, _however_ the prior CI build had actual compilation failures, so we still need to wait for a healthy windows CI build before merging.

---

Reviewed 1 of 4 files at r2, 2 of 3 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[drake/solvers/Constraint.h, line 128 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGhMZhlJuZ3mkFEX7G7:-KGhMZhmF48Kg9h4H4v7:-1816653624) ([raw file](https://github.com/robotlocomotion/drake/blob/8c1728fdde9bf9f1ce3d6166ca8fe575e0fcdda1/drake/solvers/Constraint.h#L128)):*
Use `override` here, not `virtual`, per cppguide.

---

*[drake/solvers/Constraint.h, line 130 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGhMydrG3m1-w_R-2vL:-KGhMydrG3m1-w_R-2vM:1072767542) ([raw file](https://github.com/robotlocomotion/drake/blob/8c1728fdde9bf9f1ce3d6166ca8fe575e0fcdda1/drake/solvers/Constraint.h#L130)):*
Do not say `virtual` when `override` is present, per cppguide.

---

*[drake/solvers/Constraint.h, line 140 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGhN2vyYjEl82KW1tKV:-KGhN2vyYjEl82KW1tKW:1072767542) ([raw file](https://github.com/robotlocomotion/drake/blob/8c1728fdde9bf9f1ce3d6166ca8fe575e0fcdda1/drake/solvers/Constraint.h#L140)):*
Do not say `virtual` when `override` is present, per cppguide.

---

*[drake/solvers/Optimization.h, line 586 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGhNwk1_kJsYq3UGWY_:-KGhNwk1_kJsYq3UGWYa:422327754) ([raw file](https://github.com/robotlocomotion/drake/blob/8c1728fdde9bf9f1ce3d6166ca8fe575e0fcdda1/drake/solvers/Optimization.h#L586)):*
There is no "vars parameter"?

---

*[drake/solvers/test/testOptimizationProblem.cpp, line 559 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGhOIyV9OSAbgsfp8Zb:-KGhOIyW9b4XMxG-VRLE:-1800046794) ([raw file](https://github.com/robotlocomotion/drake/blob/8c1728fdde9bf9f1ce3d6166ca8fe575e0fcdda1/drake/solvers/test/testOptimizationProblem.cpp#L559)):*
Missing required whitespace around `-` operators within these parentheses.

---

*[drake/util/Polynomial.h, line 202 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGhP2IlmU8M21b7AYwy:-KGhP2IlmU8M21b7AYwz:-1467254842) ([raw file](https://github.com/robotlocomotion/drake/blob/8c1728fdde9bf9f1ce3d6166ca8fe575e0fcdda1/drake/util/Polynomial.h#L202)):*
Why two separate doxygen comments (this one and the next block)?

---

*[drake/util/Polynomial.h, line 210 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGhPBuFtxDyEgeyw8uM:-KGhPBuFtxDyEgeyw8uN:-1382078837) ([raw file](https://github.com/robotlocomotion/drake/blob/8c1728fdde9bf9f1ce3d6166ca8fe575e0fcdda1/drake/util/Polynomial.h#L210)):*
Unclear why these two `pow` static methods are public.

---

*[drake/util/Polynomial.h, line 253 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGhPdv9UMnFFX4Qh_RG:-KGhPdvAj7uCJihqEoiw:-1467254842) ([raw file](https://github.com/robotlocomotion/drake/blob/8c1728fdde9bf9f1ce3d6166ca8fe575e0fcdda1/drake/util/Polynomial.h#L253)):*
Why two separate doxygen comments (this one and the next block)?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-:-KGhQE950-_BjNYS-y44:-1031177195)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Review status: 0 of 4 files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/solvers/Constraint.h, line 128 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGhMZhlJuZ3mkFEX7G7:-KGl_o_2yOoDprdtgDgE:-364052761) ([raw file](https://github.com/robotlocomotion/drake/blob/8c1728fdde9bf9f1ce3d6166ca8fe575e0fcdda1/drake/solvers/Constraint.h#L128)):*
Fixed this, and all nine other preexisting occurrences in this file for consistency.

---

*[drake/solvers/Constraint.h, line 130 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGhMydrG3m1-w_R-2vL:-KGl_sXOuRAMlTRiBEqU:-189593077) ([raw file](https://github.com/robotlocomotion/drake/blob/8c1728fdde9bf9f1ce3d6166ca8fe575e0fcdda1/drake/solvers/Constraint.h#L130)):*
See above

---

*[drake/solvers/Constraint.h, line 140 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGhN2vyYjEl82KW1tKV:-KGl_uJCObiuZZBdcgLj:-189593077) ([raw file](https://github.com/robotlocomotion/drake/blob/8c1728fdde9bf9f1ce3d6166ca8fe575e0fcdda1/drake/solvers/Constraint.h#L140)):*
See above

---

*[drake/solvers/Optimization.h, line 586 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGhNwk1_kJsYq3UGWY_:-KGla46PoIAZ6LeSup2-:1918676440) ([raw file](https://github.com/robotlocomotion/drake/blob/8c1728fdde9bf9f1ce3d6166ca8fe575e0fcdda1/drake/solvers/Optimization.h#L586)):*
Whee, copypasta.  Fixed.

---

*[drake/solvers/test/testOptimizationProblem.cpp, line 559 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGhOIyV9OSAbgsfp8Zb:-KGlaa8PtqApyhZGAqiD:-497468049) ([raw file](https://github.com/robotlocomotion/drake/blob/8c1728fdde9bf9f1ce3d6166ca8fe575e0fcdda1/drake/solvers/test/testOptimizationProblem.cpp#L559)):*
Fixed

---

*[drake/util/Polynomial.h, line 202 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGhP2IlmU8M21b7AYwy:-KGlaeV9si0Ue7QzmYzx:-1269866176) ([raw file](https://github.com/robotlocomotion/drake/blob/8c1728fdde9bf9f1ce3d6166ca8fe575e0fcdda1/drake/util/Polynomial.h#L202)):*
This is a documented way to provide a brief and verbose description and generates pretty doxygen without putting format hieroglyphs in the code.  See: https://www.stack.nl/~dimitri/doxygen/manual/docblocks.html#cppblock (from "a third option...")

---

*[drake/util/Polynomial.h, line 210 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGhPBuFtxDyEgeyw8uM:-KGlb4ZNa8-j-kKNLsdQ:1758359711) ([raw file](https://github.com/robotlocomotion/drake/blob/8c1728fdde9bf9f1ce3d6166ca8fe575e0fcdda1/drake/util/Polynomial.h#L210)):*
Fixed.

---

*[drake/util/Polynomial.h, line 253 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGhPdv9UMnFFX4Qh_RG:-KGlbQQ09id3gBYreke0:1133110569) ([raw file](https://github.com/robotlocomotion/drake/blob/8c1728fdde9bf9f1ce3d6166ca8fe575e0fcdda1/drake/util/Polynomial.h#L253)):*
See previous response:  This is a documented form of doxygen comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2215)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 4 of 4 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/solvers/Constraint.h, line 128 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGhMZhlJuZ3mkFEX7G7:-KGlfftbChg5RRqhgUGj:182093779) ([raw file](https://github.com/robotlocomotion/drake/blob/8c1728fdde9bf9f1ce3d6166ca8fe575e0fcdda1/drake/solvers/Constraint.h#L128)):*
It looks like the destructor still says virtual and not override?

---

*[drake/util/Polynomial.h, line 202 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGhP2IlmU8M21b7AYwy:-KGlfwm-Owu2tqTdXoqd:733624448) ([raw file](https://github.com/robotlocomotion/drake/blob/8c1728fdde9bf9f1ce3d6166ca8fe575e0fcdda1/drake/util/Polynomial.h#L202)):*
Wow, TIL.  I guess I wish we used autobrief, but whatever, this PR is fine.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2215)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Review status: 3 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/solvers/Constraint.h, line 128 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGhMZhlJuZ3mkFEX7G7:-KGlihqkBWY74sT_J2kc:999278321) ([raw file](https://github.com/robotlocomotion/drake/blob/8c1728fdde9bf9f1ce3d6166ca8fe575e0fcdda1/drake/solvers/Constraint.h#L128)):*
Sure enough.  Fixed.  (and the other nine as well).

---

*[drake/util/Polynomial.h, line 202 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-KGhP2IlmU8M21b7AYwy:-KGlirWYWOUMX1GWZHPc:1420744282) ([raw file](https://github.com/robotlocomotion/drake/blob/8c1728fdde9bf9f1ce3d6166ca8fe575e0fcdda1/drake/util/Polynomial.h#L202)):*
I'm just happy to have something that isn't the weird repeat-the-name-then-`\brief`convention here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2215)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-:-KGls26KrTLPV6HLBRZ2:1417088721)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2215#-:-KGlsHBl3aLd45EzzcSD:-1868698381)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Ready to merge iff jenkins passes.

ggould-tri commented:
Drake005 failure seems to be rooted at "execvp error on file /tmp/drake-2215-27705/build/bin/bertini (No such file or directory)" which probably is not me.

Windows failure seems to be some sort of C preprocessor fail, which maybe is me.  Investigating.
